python/tritonserver/_api/_allocators.py

```diff
@@ -0,0 +1,323 @@
+        memory_type_id : int
+            memory type id requested (typically device id)
+        tensor_name : str
+            name of tensor
+
+        Returns
+        -------
+        MemoryBuffer
+            memory buffer with requested size
+
+        Examples
+        --------
+        memory_buffer = allocator.allocate(100,MemoryType.CPU,0,"output")
+
+        """
+
+        pass
+
+
+class NumpyAllocator(MemoryAllocator):
+    def __init__(self):
+        pass
+
+    def allocate(
+        self, size: int, memory_type: MemoryType, memory_type_id: int, tensor_name: str
+    ) -> MemoryBuffer:
+        ndarray = numpy.empty(size, numpy.byte)
+        return MemoryBuffer.from_dlpack(ndarray)
+
+
+default_memory_allocators[MemoryType.CPU] = NumpyAllocator()
+
+if cupy is not None:
+
+    class CupyAllocator(MemoryAllocator):
+        def __init__(self):
+            pass
+
+        def allocate(
+            self,
+            size: int,
+            memory_type: MemoryType,
+            memory_type_id: int,
+            tensor_name: str,
+        ) -> MemoryBuffer:
+            with cupy.cuda.Device(memory_type_id):
+                ndarray = cupy.empty(size, cupy.byte)
+
+            return MemoryBuffer.from_dlpack(ndarray)
+
+    default_memory_allocators[MemoryType.GPU] = CupyAllocator()
+
+
+class ResponseAllocator:
+    def __init__(
+        self,
+        memory_allocator: Optional[MemoryAllocator] = None,
+        device_or_memory_type: Optional[DeviceOrMemoryType] = None,
+    ):
+        self._memory_allocator = memory_allocator
+        self._memory_type: Optional[MemoryType] = None
+        self._memory_type_id: int = 0
+        self._response_allocator = None
+        if device_or_memory_type is not None:
+            self._memory_type, self._memory_type_id = parse_device_or_memory_type(
+                device_or_memory_type
+            )
+
+    def allocate(
+        self,
+        _allocator,
+        tensor_name,
+        byte_size,
+        memory_type,
+        memory_type_id,
+        _user_object,
+    ):
+        if self._memory_type is not None:
+            memory_type = self._memory_type
+            memory_type_id = self._memory_type_id
+
+        memory_allocator = self._memory_allocator
+        if memory_allocator is None:
+            memory_allocator = default_memory_allocators[memory_type]
+
+        memory_buffer = memory_allocator.allocate(
+            byte_size, memory_type, memory_type_id, tensor_name
+        )
+
+        return (
+            memory_buffer.data_ptr,
+            memory_buffer,
+            memory_buffer.memory_type,
+            memory_buffer.memory_type_id,
+        )
+
+    def release(
+        self,
+        _allocator,
+        _buffer_,
+        _buffer_user_object,
+        _byte_size,
+        _memory_type,
+        _memory_type_id,
+    ):
+        pass
+
+    def start(self, _allocator, _user_object):
+        pass
```

Comment:
what does it mean to "start" an allocator?

Reply:
this is part of the c api - it is to indicate a start of allocation for a request - decided not to expose to public allocator interface but can add it in the future.